rust/src/composepost.rs

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2018 Red Hat, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+use openat;
+use std::{io, fs};
+use std::io::{Write, BufRead};
+use std::path::Path;
+use failure::Fallible;
+
+// IMO should propose this at least in a "utils" bit of the openat crate;
+// Like 95% of the time I'm looking at errno (with files) it's for ENOENT,
+// and Rust has an elegant way to map that with Option<>.  Every other
+// error I usually just want to propagate back up.
+fn openat_optional<P, F>(d: &openat::Dir, p: P, f: F) -> io::Result<Option<fs::File>>
+    where P: openat::AsPath, F: FnOnce(&openat::Dir, P) -> io::Result<fs::File>
+{
+    match f(d, p) {
+        Ok(f) => Ok(Some(f)),
+        Err(e) => {
+            if e.kind() == io::ErrorKind::NotFound {
+                Ok(None)
+            } else {
+                Err(e)
+            }
+        }
+    }
+}
+
+// rpm-ostree uses /home → /var/home by default as generated by our
+// rootfs; we don't expect people to change this.  Let's be nice
+// and also fixup the $HOME entries generated by `useradd` so
+// that `~` shows up as expected in shells, etc.
+//
+// https://github.com/coreos/fedora-coreos-config/pull/18
+// https://pagure.io/workstation-ostree-config/pull-request/121
+// https://discussion.fedoraproject.org/t/adapting-user-home-in-etc-passwd/487/6
+// https://github.com/justjanne/powerline-go/issues/94
+fn postprocess_useradd(rootfs_dfd: &openat::Dir) -> Fallible<()>
+{
+    let path = Path::new("usr/etc/default/useradd");
+    if let Some(f) = openat_optional(rootfs_dfd, path, openat::Dir::open_file)? {
```

Comment:
Minor: we can avoid the indentation here by returning early in the `None` case.

Reply:
I used to have an early return here but it feels like...bad style?   We don't have a huge amount of indentation here anyways.